pkg/apiserver/proxy.go

```diff
@@ -152,14 +153,51 @@ func (r *ProxyHandler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
 	}
 	newReq.Header = req.Header
 
-	proxy := httputil.NewSingleHostReverseProxy(&url.URL{Scheme: "http", Host: destURL.Host})
-	proxy.Transport = &proxyTransport{
-		proxyScheme:      req.URL.Scheme,
-		proxyHost:        req.URL.Host,
-		proxyPathPrepend: path.Join(r.prefix, "ns", namespace, kind, id),
+	if req.Header.Get("Connection") == "Upgrade" {
+		kubeletConn, err := net.Dial("tcp", destURL.Host)
+		if err != nil {
+			status := errToAPIStatus(err)
+			writeJSON(status.Code, r.codec, status, w)
+			return
+		}
+
+		requestHijackedConn, _, err := w.(http.Hijacker).Hijack()
+		if err != nil {
+			status := errToAPIStatus(err)
+			writeJSON(status.Code, r.codec, status, w)
+			return
+		}
+
+		defer requestHijackedConn.Close()
+		defer kubeletConn.Close()
+
+		err = newReq.Write(kubeletConn)
+		if err != nil {
+			status := errToAPIStatus(err)
+			writeJSON(status.Code, r.codec, status, w)
+			return
+		}
+
+		errc := make(chan error, 2)
+		cp := func(dst io.Writer, src io.Reader) {
+			_, err := io.Copy(dst, src)
+			errc <- err
+		}
+		go cp(kubeletConn, requestHijackedConn)
```

Comment:
Ah, nm, I see you are blocking on those error channels.  Wouldn't a wait group make more sense for this?